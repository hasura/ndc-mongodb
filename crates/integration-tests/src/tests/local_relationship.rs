use crate::{connector::Connector, graphql_query, run_connector_query};
use insta::assert_yaml_snapshot;
use ndc_test_helpers::{asc, field, query, query_request, relation_field, relationship};

#[tokio::test]
async fn joins_local_relationships() -> anyhow::Result<()> {
    assert_yaml_snapshot!(
        graphql_query(
            r#"
                query {
                  movies(limit: 2, order_by: {title: Asc}, where: {title: {_iregex: "Rear"}}) {
                    id
                    title
                    comments(limit: 2, order_by: {id: Asc}) {
                      email
                      text
                      movie {
                        id
                        title
                      }
                      user {
                        email
                        comments(limit: 2, order_by: {id: Asc}) {
                          email
                          text
                          user {
                            email
                            comments(limit: 2, order_by: {id: Asc}) {
                              id
                              email
                            }
                          }
                        }
                      }
                    }
                  }
                }
            "#
        )
        .run()
        .await?
    );
    Ok(())
}

#[tokio::test]
async fn filters_by_field_of_related_collection() -> anyhow::Result<()> {
    assert_yaml_snapshot!(
        graphql_query(
            r#"
            query {
              comments(where: {movie: {rated: {_eq: "G"}}}, limit: 10, order_by: {id: Asc}) {
                movie {
                  title
                  year
                }
              }
            }
            "#
        )
        .run()
        .await?
    );
    Ok(())
}

#[tokio::test]
async fn filters_by_non_null_field_of_related_collection() -> anyhow::Result<()> {
    assert_yaml_snapshot!(
        graphql_query(
            r#"
            query {
              comments(
                limit: 10
                where: {movie: {title: {_is_null: false}}}
                order_by: {id: Asc}
              ) {
                movie {
                  title
                  year
                }
              }
            }
            "#
        )
        .run()
        .await?
    );
    Ok(())
}

#[tokio::test]
async fn filters_by_field_of_relationship_of_relationship() -> anyhow::Result<()> {
    assert_yaml_snapshot!(
        graphql_query(
            r#"
            query {
              artist(where: {albums: {tracks: {name: {_eq: "Princess of the Dawn"}}}}) {
                name
                albums(order_by: {title: Asc}) {
                  title
                }
              }
            }
            "#
        )
        .run()
        .await?
    );
    Ok(())
}

#[tokio::test]
async fn sorts_by_field_of_related_collection() -> anyhow::Result<()> {
    // Filter by rating to filter out comments whose movie relation is null.
    assert_yaml_snapshot!(
        graphql_query(
            r#"
            query {
              comments(
                limit: 10
                order_by: [{movie: {title: Asc}}, {date: Asc}]
                where: {movie: {rated: {_eq: "G"}}}
              ) {
                movie {
                  title
                  year
                }
                text
              }
            }
            "#
        )
        .run()
        .await?
    );
    Ok(())
}

#[tokio::test]
async fn looks_up_the_same_relation_twice_with_different_fields() -> anyhow::Result<()> {
    assert_yaml_snapshot!(
        graphql_query(
            r#"
            query {
              artist(limit: 2, order_by: {id: Asc}) {
                albums1: albums(order_by: {title: Asc}) {
                  title
                }
                albums2: albums(order_by: {title: Asc}) {
                  tracks(order_by: {name: Asc}) {
                    name
                  }
                }
              }
            }
            "#
        )
        .run()
        .await?
    );
    Ok(())
}

#[tokio::test]
async fn queries_through_relationship_with_null_value() -> anyhow::Result<()> {
    assert_yaml_snapshot!(
        graphql_query(
            r#"
            query {
              comments(where: {id: {_eq: "5a9427648b0beebeb69579cc"}}) { # this comment does not have a matching movie
                movie {
                  comments {
                    email
                  }
                } 
              }
            }
            "#
        )
        .run()
        .await?
    );
    Ok(())
}

#[tokio::test]
async fn joins_on_field_names_that_require_escaping() -> anyhow::Result<()> {
    assert_yaml_snapshot!(
        run_connector_query(
            Connector::TestCases,
            query_request()
                .collection("weird_field_names")
                .query(
                    query()
                        .fields([
                            field!("invalid_name" => "$invalid.name"),
                            relation_field!("join" => "join", query().fields([
                              field!("invalid_name" => "$invalid.name")
                            ]))
                        ])
                        .order_by([asc!("_id")])
                )
                .relationships([(
                    "join",
                    relationship("weird_field_names", [("$invalid.name", "$invalid.name")])
                )])
        )
        .await?
    );
    Ok(())
}
