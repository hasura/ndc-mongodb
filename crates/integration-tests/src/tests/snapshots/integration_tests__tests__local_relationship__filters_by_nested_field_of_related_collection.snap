---
source: crates/integration-tests/src/tests/local_relationship.rs
expression: "graphql_query(r#\"\n            query {\n              comments(where: {movie: {imdb: {rating: {_eq: 9.2}}}}, limit: 10, order_by: {id: Asc}) {\n                text\n                movie {\n                  title\n                  year\n                }\n              }\n            }\n            \"#).run().await?"
---
data:
  comments:
    - text: Odit libero reprehenderit vitae dolore. Consequuntur neque ab eos id quibusdam. Odit laborum ex rem.
      movie:
        title: The Godfather
        year: 1972
    - text: Minima cupiditate veritatis rerum sequi vitae illum. Ea eos voluptas non adipisci quidem suscipit cum quo. In itaque cum architecto non blanditiis.
      movie:
        title: The Godfather
        year: 1972
    - text: Enim id inventore magnam hic. Error a nobis earum delectus saepe qui. Tenetur odit facilis ratione voluptates. Aliquid ullam temporibus porro mollitia.
      movie:
        title: The Godfather
        year: 1972
    - text: Eveniet aliquid et quae consectetur saepe repellendus veritatis voluptates. Harum porro facilis modi in. Dolor architecto vel reiciendis fuga.
      movie:
        title: The Godfather
        year: 1972
    - text: Aliquid iste eos facere corporis repellendus doloremque. Beatae libero itaque dolorum delectus debitis. Pariatur optio tempora pariatur unde culpa dolores.
      movie:
        title: The Godfather
        year: 1972
    - text: Doloribus esse quos atque voluptatem nihil libero molestias porro. Accusantium sapiente perferendis quae fugiat dicta. Ab consequuntur dicta blanditiis enim vitae odit tempora.
      movie:
        title: The Godfather
        year: 1972
    - text: Unde ratione ad nemo eligendi cupiditate aut iusto natus. Consequuntur cum totam laborum rerum dolorum sint. Odit recusandae autem dolorum natus sunt.
      movie:
        title: The Godfather
        year: 1972
    - text: Veniam doloribus ea cum error. Voluptates voluptatum tempore culpa earum nihil. Voluptate ad asperiores architecto hic.
      movie:
        title: The Godfather
        year: 1972
    - text: Eveniet voluptate quo quas quia fuga eos quibusdam. Neque molestiae quis assumenda nostrum. Ipsa laudantium esse voluptas recusandae.
      movie:
        title: The Godfather
        year: 1972
    - text: Mollitia ut sint veniam accusamus quam. Quod ullam voluptatum perspiciatis odio. Non ullam maxime recusandae ipsam possimus sit. Ullam doloribus quisquam at.
      movie:
        title: The Godfather
        year: 1972
errors: ~
